File: app/src/main/java/com/stho/software/nyota/utilities/Location.java
package com.stho.software.nyota.utilities;

/**
 * GEO-Location with latitude in degrees, longitude in degrees, altitude in km
 */
public class Location implements ILocation {

    private final double latitude; // in degrees
    private final double longitude; // in degrees
    private final double altitude; // in km

    private final static double ONE_METER_IN_KM = 0.001;

    /**
     * Define a new GEO-location. Altitude is set to 0.
     * @param latitude in degrees
     * @param longitude in degrees.
     */
    Location(double latitude, double longitude) {
        this.latitude = latitude;
        this.longitude = longitude;
        this.altitude = 0;
    }

    /**
     * Define a new GEO-location.
     * @param latitude in degrees
     * @param longitude in degrees
     * @param altitude in km
     */
    public Location(double latitude, double longitude, double altitude) {
        this.latitude = latitude;
        this.longitude = longitude;
        this.altitude = altitude;
    }

    Location(android.location.Location location) {
        this(location.getLatitude(), location.getLongitude(), ONE_METER_IN_KM * location.getAltitude());
    }

    @Override
    public String toString() {
        return Location.toString(getLatitude(), getLongitude());
    }

    public static String toString(android.location.Location location) {
        return Location.toString(location.getLatitude(), location.getLongitude());
    }

    public static String toString(double latitude, double longitude) {
        return Angle.toString(longitude, Angle.AngleType.LONGITUDE) + Formatter.SPACE + Angle.toString(latitude, Angle.AngleType.LATITUDE);
    }

    @Override
    public double getLatitude() {
        return latitude;
    }

    @Override
    public double getLongitude() {
        return longitude;
    }

    @Override
    public double getAltitude() {
        return altitude;
    }
}

File: app/src/main/java/com/stho/nyota/sky/utilities/Location.kt
package com.stho.nyota.sky.utilities

import com.stho.nyota.sky.universe.Algorithms
import com.stho.nyota.sky.universe.altitudeInKm
import org.osmdroid.util.GeoPoint


/**
 * GEO-Location with latitude in degrees, longitude in degrees, altitude in km
 *
 * Define a new GEO-location.
 * @param latitude in degrees
 * @param longitude in degrees
 * @param altitude in km
 */
class Location(override val latitude: Double, override val longitude: Double, override val altitude: Double = 0.0) : ILocation {

    override fun toString(): String =
        toString(latitude, longitude)

    internal fun toGeoPoint(): GeoPoint =
        GeoPoint(latitude, longitude, altitude)

    fun isNearTo(otherLocation: ILocation): Boolean =
        getHorizontalDistanceInKmTo(otherLocation) < TEN_METERS_IN_KM

    fun getHorizontalDistanceInKmTo(otherLocation: ILocation): Double =
        Algorithms.getHorizontalDistanceInKmTo(latitude, longitude, otherLocation.latitude, otherLocation.longitude)

    companion object {

        private const val TEN_METERS_IN_KM = 0.01

        internal fun fromAndroidLocation(location: android.location.Location): Location =
            Location(location.latitude, location.longitude, location.altitudeInKm())

        internal fun fromGeoPoint(geoPoint: GeoPoint): Location =
            Location(geoPoint.latitude, geoPoint.longitude, geoPoint.altitude)

        fun toString(location: android.location.Location): String =
            toString(location.latitude, location.longitude)

        fun toString(latitude: Double, longitude: Double): String =
            Angle.toString(longitude, Angle.AngleType.LONGITUDE) + Formatter.SPACE + Angle.toString(latitude, Angle.AngleType.LATITUDE)

        fun getDefault(): Location =
            Location(0.0,0.0)
    }
}